### ISSUE TITLE
Bitwise Operators

### ISSUE THREAD

**Issue by Redrield** [Kotlin/KEEP#142]
This is the issue for discussion to the proposal of adding [overloadable bitwise operators](https://github.com/Redrield/KEEP/blob/bitwise-operators/proposals/bitwise-operators.md) into Kotlin

Proposal PR: #143 

Prior discussion about the merits/drawbacks of bitwise operators can be found at [KT-1440](https://youtrack.jetbrains.com/issue/KT-1440)

**Comment by ilya-g** [comment]
The proposed operator names are inconsistent, usually for operator `op` there is `opAssign` assignment counterpart.

I wonder what is the reason for introducing new names for these operators? Have you considered an option of just adding the `operator` modifier to the existing `or`, `and`, `xor` functions?

**Comment by Redrield** [comment]
@ilya-g To my knowledge it is not possible for a function to be both `operator` and `infix`. If I were to modify the existing infix functions, that would break compatibility. Which is something I would not like to do with this proposal. The names may be subject to change, when I was creating the prototype (Written prior to this proposal), I was using the first names that came to my head.

**Comment by ilya-g** [comment]
> it is not possible for a function to be both operator and infix

I don't remember such limitation. For example it is possible to declare an operator `plus` both as operator and infix function:

```
object Zero {
    infix operator fun plus(other: Zero): Zero = this
}
```

**Comment by Redrield** [comment]
I've updated the proposal as such in commit 80e9e3e. I was not aware that `infix` and `operator` were not mutually exclusive.

**Comment by elect86** [comment]
Also, the bitwise inversion `~`

**Comment by Wasabi375** [comment]
While I somewhat like the idea of adding the binary operators I'm not sure that they are necessary.
I would love to see a better solution for `a = a or b`, but we also need to consider that binary operations are used rarely in most programs. AFAIK they are only(mostly) used to interface with native C code as C libraries often use bit fields to represent a group of booleans.
The question IMO is whether those few programs are important enough to add a few more operators. Except for the effort of implementing them there are not really many downsides (I have no idea how much this would be):
1. Those operators will probably be used for other operations, but than I'm not sure how often operators are used for other operations than their mathematical function already. If this really is an issue it is already with the other operators. Also there is no way of stopping people writing bad code.
2.  many people don't really know those operators or don't know them well.
`shr`, `ushr` are fairly self explanatory but how many people can tell you which one is `>>` and which one is `>>>`. I have to look this up every time I work with bits in java. They are just not used often enough. And yes I can see that people who know nothing about bitwise operators won't understand `shr` either but they will have to look up the operator as well.
3. They are simply not used often

Arguments for bitwise operators are:
1. As stated in the KEEP, operator precedence is missing for infix functions. But than again longer bitwise operations should use parentheses anyways as they get confusing otherwise. 
2. `a |= b`, etc is much easier to read than `a = a or b`
3. Most languages support those operators, which will make switching languages easier. I'm not sure though about this argument. Only because something is the norm does not mean it's the right decision. Plus the familiarity argument is less impactful with something like this which is not used often. 

___
I think the KEEP should reference the issue about this: https://youtrack.jetbrains.com/issue/KT-1440

**Comment by altavir** [comment]
I agree with @Wasabi375. I work a lot with binary data and even in my case bit-wise operations are used very rarely. In most cases the bit-wise operation is a possible source of errors, so those calls should be encapsulated in separate functions. 

In my opinion the possibility of someone using those operations not the way they are intended is high. For example I have experience with groovy, where bit-wise shifts are used for some IO operations like in C++ and it is confusing as hell.

**Comment by Redrield** [comment]
@altavir Part of that problem is that the standard libraries for groovy and C++ include overloads of operators that are used outside of bitwise operations. In this case the kotlin stdlib will only provide overloads of these operators on numerics, doing as you would expect. I don't think that someone writing unmaintainable code for themselves should be a stopping factor in this.

**Comment by Dico200** [comment]
From the file:
* Should the functions be marked non-`infix` in the future?
* How would a deprecation notice be shown to notify users of the deprecated status of calling the functions in `infix` notation

Deprecating `infix` calls would usually imply that the feature is intended to be removed.
I think it would be incredibly tedious to remove `infix` modifier from the currently available functions. Introducing compiler errors into previously fine source code is not something to take lightly.
In addition, there isn't a way to express deprecation of calls in infix notation other than implementing that into the compiler (correct me if I'm wrong).

Not removing the `infix` modifier means there are arithmetic-performing operator functions like `plus` that are not `infix`. These could easily be made `infix` for consistency, without breaking any backwards compatible constraints. In addition, it makes sense that you could write `8 plus 3` if you could write `8 shr 3`. I cannot think of other compelling reasons to do this.

Doing this means no deprecation of any existing functions is required. You can call the functions with regular function-call-with-receiver syntax anyway.

**Comment by Mishkun** [comment]
@Redrield Have you thought about using existing Boolean operators `&&`, `||`, `!` in conjunction with inline classes released in 1.3? You could make the `BitwiseInt` inline wrapper and overload operators for it. 
I'm huge fan of operator overloading and defining custom operators with custom precedence and associativity, but I'd rather be the _all-or-nothing_ guy.

**Comment by Redrield** [comment]
@Mishkun Logical AND and OR have lower precedence than every one of the operators proposed in this document. It is also grossly inconsistent to have && take the role of both logical and bitwise AND. Not only that, such a solution also leaves no way to represent XOR, and bitshift operations with operators.

**Comment by Dico200** [comment]
@Mishkun it is actually not possible to overload && and II operators. I just tried, the declaration does not exist. For `Boolean`, the functions are `infix`. It must be compiler intrinsic.

**Comment by Dico200** [comment]
@Redrield If you do want to use the symbols we know as binary operators, you can name your infix functions `\`$\`` where $ is the symbol.

**Comment by Redrield** [comment]
@Dico200 That may work as a workaround but that still runs into the issue of infix operators having no precedence.

**Comment by Dico200** [comment]
> Also despite looking similar, the fact that it would be `` a `&` b `` instead of `a & b` would still be confusing to people new to Kotlin.

@Redrield Sorry, I did not suggest that this should be implemented into the standard library. I only suggested a workaround you can add to your own code.

**Comment by Mishkun** [comment]
@Dico200 
> it is actually not possible to overload && and II operators
Oh, that's a twist.

**Comment by anikaiful** [comment]
> The proposed operator names are inconsistent, usually for operator `op` there is `opAssign` assignment counterpart.
> 
> I wonder what is the reason for introducing new names for these operators? Have you considered an option of just adding the `operator` modifier to the existing `or`, `and`, `xor` functions?
Current 'or' and 'and' operators are confusing and illogical.

**Comment by anikaiful** [comment]
> While I somewhat like the idea of adding the binary operators I'm not sure that they are necessary.
> I would love to see a better solution for `a = a or b`, but we also need to consider that binary operations are used rarely in most programs. AFAIK they are only(mostly) used to interface with native C code as C libraries often use bit fields to represent a group of booleans.
Or any application which relies on bitwise flags.

**Comment by Mishkun** [comment]
> Or any application which relies on bitwise flags. Such approach is by far not uncommon. And which is why I'd just LOVE a way to and/or ENUM's, akin how C# does it.
In Java enums are very different from C#. In Java enums are objects and in C# enums are just fancy integers.

**Comment by Mishkun** [comment]
You can make your own MultiEnum<T: Enum> class backed by Set and overload `contains` operator on that. And then you can make some factory functions and maybe even overload some binary ops.
But this conversation is out of scope of this proposal

**Comment by elect86** [comment]
I think that the argument "no bitwise operator otherwise some people will abuse it" is very wrong and dangerous

One can already today write very bad and unreadable code, but not for this reason a lot of cool features in Kotlin are forbidden.

We shall teach developers writing nice code because it improves readability, which in turns make it easier to understand, lesser error prone, quicker to solve bugs, etc etc.. 

Forbidding a feature because sometimes somebody will abuse it, I think it's simply illogical and silly

**Comment by gildor** [comment]
> improves readability, which in turns make it easier to understand, lesser error prone, quicker to solve bugs

@elect86 You probably talk not about bitwise operators, because their use case (even good use cases) are completely opposite: hard to understand, very error prone, hard to find bugs. 
I don't tell that infix functions are better in this case, I just point that bit fiddling, in general, is hard and error-prone.
You can even check "Use cases" section of this proposal and see how bad it is with or without flags.
Even simple (and real life) use case could be written without bitwise operators or infix functions:
``` someView.setGravity(Gravity.BOTTOM, Gravity.LEFT)```
No any sense to use the mask in this particular case, even if you can in some cases combine them. For such use cases, EnumSet works perfectly as well.

And this sample is just insane, with or without bitwise operators
```
flags = (flags & ~FLAG_A) | FLAG_B | FLAG_C
```

I'm pretty sure there are use cases when it makes sense, but it's completely non-readable, error-prone and hard to follow.
I would just rewrite it (with something like enum set or just an API) to:
```
someClass
   .addFlag(FLAB_B, FLAG_C)
   .removeFlag(FLAG_A)
```

So much more readable and easy to use.

> There are many Android APIs where the parameters to a method for some operation will contain int flags, which Kotlin currently does not handle as gracefully as it could due to the infix functions replacing these common operators.

Most of such use cases even for existing APIs could be solved by extension function that make such API type safe, easy to use and provide better discoverability.

I want to say again, I know there are areas when work with bits is much more common (graphics, system programming and so on), so because of this I support idea of bitwise operators.

But for now everyone talks about flags what is sounds strange for me, especially when you have all possibilities of Kotlin extension functions.

Also, question about bitwise operators overloading is still open for me. I think that if you introduce them, you cannot avoid overloading, for example now we have unsigned numbers API and you need it there. But problem with abuse of bit operators (and proposal contains a lot of them, especially infamous (in terms of oveloading) `>>` and `<<`) is still big problem of this proposal

**Comment by Dico200** [comment]
There are numerous examples where ints are used in Java libraries where they might as well be unsigned. Many of them are an application of bitwise logic. Since unsigned types can't be expressed in Java, similar to the lack of ability to express nullability of a type, maybe we can introduce some annotations that allow accessing java `int` with kotlin `UInt`, for example.

If we do this, we can justify dropping the majority of bitwise operators for signed numbers. `>>>` for example will be obsolete.

Something else to consider would be to identify common patterns in bitwise logic. Maybe we can come up with a new system that is much easier to use. For example, to remove bits from another number, one commonly writes `a & ~b` but this is quite similar to `-`, in which case `|` is similar to `+`. Difference is that there isn't always a change with the bitwise version, but it would be much easier to read and write.

Another common operation is acquiring a mask of x bits long. In java, a weird quirk you can use is `-1 >>> -x` which will give you `x` set bits on the least significant side, for reasons I don't really get. I think we should avoid strange behaviour like this. How can you express width of a mask to create? Maybe something with `[]`, would make sense. Maybe we can use the `..` operator: bit x to bit y. 

Of course, we must evaluate that all things you can do with the bitwise operators can be achieved with this potential new "system".

**Comment by elect86** [comment]
just wasted half an hour debugging because of `shl` didn't follow the same order priority as `<<`

`memAddress(buffer) + pos shl 2`

Another good reason

**Comment by gildor** [comment]
Even this code `memAddress(buffer) + pos << 2` requires knowledge about operation order (same with infix functions), so If I would see this on code review would suggest using square brackets or intermediate variable

**Comment by Redrield** [comment]
For the example of EnumSet APIs that was brought up by @gildor, I've been doing a lot of work with Rust where a library `bitflags` provides something similar, with operator overloading.

```rust
 bitflags! {
    struct Flags: u8 {
        const A = 0b0000_0001;
        const B = 0b1000_0000;
    }
}
```
is the API that is given. Pulling this apart, it creates a typesafe wrapper around the values that you specify, with A, B (and however many other arbitrary constants) being instances of a struct with the bit value specified. The struct that wraps them also provides overloads for bitwise operations. This sort of type-safe flag system is one that could be provided in Kotlin, and would be the best of both worlds.

(Here's some example usage of bitflags in rust) 
```rust
        let mut control = self.mode.to_control();

        if self.enabled {
            control |= Control::ENABLED;
        }

        if self.estopped {
            control |= Control::ESTOP
        }

```

**Comment by Redrield** [comment]
A lot of time has passed since the last activity on this post. I've updated the proposal with some more details around the implementation (or at least how I implemented my prototype in the compiler), as well as the list of classes that would receive implementations of these operators in the standard library, if they were to be added.

I would prefer that instead of allowing the conversation around this to stagnate, a discussion could be had to reach a final decision over whether or not this proposal will be accepted for a future version of Kotlin. It's clear there are arguments on both sides of this issue, however not providing a concrete answer as to the future of this proposal won't leave either side happy.

**Comment by Dico200** [comment]
@gildor said:
> Also, the question about bitwise operators overloading is still open for me. I think that if you introduce them, you cannot avoid overloading, for example now we have unsigned numbers API and you need it there.

I think an easy solution would be to restrict  certain operators to the standard library, or something like that, with a compiler intrinsic. Perhaps this would be a good idea for bitshift operators.

**Comment by Redrield** [comment]
I understand the concern with the history of abuse of bitwise operator overloading in particular in several other languages, however disallowing overloading of these operators also disallows a typesafe flag API such as the one I mentioned above. Being able to retain familiar code practices while also guaranteeing the validity of flags ahead of time is a powerful feature that Kotlin could have with operator overloading.

One solution that may work, is to initially add bitwise operators to Kotlin, disallow overloading and only implement them on the types mentioned in the proposal, while further discussion can be carried out as to whether they should be changed to allow universal overloading in the future.

**Comment by Wasabi375** [comment]
I don't think restricting overloading works. As long as the standard library does not abuse the operators (like C++ with `<<`) I don't think there will be much abuse of the system.  
Also we should think about libraries which have legitimate reasons to overload bitwise operators e.g the libs connected to https://github.com/mipt-npm/KEEP-math.

**Comment by altavir** [comment]
The only valid place in mathematics I can think of is the set theory, but it is really rarely used in numerical mathematics an even in those cases infix functions are quite sufficient.

I do not think that if those operators are introduced, they be ever use as actual logical operators. For example, people will immediately start to use `|` as some kind of bash pipe. So if the question is whether we need logical operators to describe some kind of valid logical operations, then probably no, we don't. If we just want few more operator to give them our own meaning and play around - it is a C++/Scala way.

Mathematical KEEP is sadly not active. Kmath seems to be the only kotlin mathematics librrary in active development right now (save for kotln-statistics), so people prefer to write short issues to its tracker instead of specifications.

**Comment by Redrield** [comment]
As was said by @Wasabi375, so long as a precedent is set by stdlib w.r.t in what circumstances these operators should use, I don't think that abuse is as likely as is made out to be.

For instance, Rust allows overloadable bitwise operators by means of traits defined in `std::ops`, and I can't think of a situation where they have been abused in the way that you may see in C++.

**Comment by altavir** [comment]
The question is not about why it is not bad, but why it is need. In order to get the feature, you need to overcome -N points rule. So first you need to show some places, where those operators are needed. For not I do not see any other than bit-wise operators and set therory ( the later works pretty good with +/- or infix functions.

**Comment by Wasabi375** [comment]
Normally I agree with the -N points rule but in this case it's the
other way around. The aim of the rule is to prevent a feature creep that
makes the language more complicated than it needs to be. But in this case
not allowing operator overloading is more complicated since it is allowed
for all other operators.

The need for bitwise operators should exceed the -N point rule to add them
at all but in that case we should also allow operator overloading for
them.

**Comment by CommanderTvis** [comment]
It would be nice:
```kotlin
operator fun PrintStream.shl(object: Any) = apply { print(object) }

System.out << "hi!" << System.lineSeparator()
```

**Comment by Redrield** [comment]
That abuse is precisely why bitwise operators are so contentious. My desire to get bitwise operators in the language is for operations that involve binary logic in a way that the ubiquity of the operators would be an improvement for the developer.

**Comment by neworld** [comment]
Current operators are abused already. For example, [official kotlinx.html](https://github.com/Kotlin/kotlinx.html#dom) uses `+` operator for texts:
```
        div {
            +"added it"
        }
```

Maybe more operators will be beneficial for such nice DSL as well?

**Comment by CommanderTvis** [comment]
> That abuse is precisely why bitwise operators are so contentious. My desire to get bitwise operators in the language is for operations that involve binary logic in a way that the ubiquity of the operators would be an improvement for the developer.

I just replicated familiar for me C++ syntax. Regular industrial development doesn't often require bitwise operations, so they can be a syntax feature like actual overload able operators.

**Comment by altavir** [comment]
@neworld yes it is abused and not everyone is happy with it. I am not sure it is a good solution in the long run.

@CommanderTvis well, it is one of the ugliest things in C++. I think that you just stated the main argument NOT to bring custom shift operators. Especially, remembering that kotlin has string interpolation.

**Comment by CommanderTvis** [comment]
> @neworld yes it is abused and not everyone is happy with it. I am not sure it is a good solution in the long run.
> 
> @CommanderTvis well, it is one of the ugliest things in C++. I think that you just stated the main argument NOT to bring custom shift operators. Especially, remembering that kotlin has string interpolation.
1. Operator abusing is promoted even by Kotlin's documentation, and it will be very strange to introduce new operators and make their overloading an antipattern.
2. I will not argue about ugliness of C++ standard library.

**Comment by neworld** [comment]
> yes it is abused and not everyone is happy with it. I am not sure it is a good solution in the long run.

I had seen many cases of abuse in well-know libraries. For example [this string invocation in Gradle Kotlin dsl](https://github.com/gradle/kotlin-dsl-samples/blob/master/samples/ant/build.gradle.kts#L18-L23)

```kotlin
                "zip"("destfile" to "$buildDir/archive.zip") {
                    "fileset"("dir" to "src") {
                        "include"("name" to "**/*.xml")
                        "include"("name" to "**/*.txt")
                    }
                }
```

The reason for this is pretty strong. As a longtime user, I don't see a problem here even my first insight was very skeptical. 

I just hope bitwise operators are not implemented due to stronger reasons than plain fear.

**Comment by CommanderTvis** [comment]
@neworld  as I know, currently, Gradle creates accessors for dynamic stuff automatically.

**Comment by gildor** [comment]
@neworld We discussing here bitwise operators particularly, not invoke operator. Invoke operator indeed very powerful and have very good use cases for DSL, please, give some examples for bitwise operators

**Comment by altavir** [comment]
I am in fact not against additional custom operations yo use in builders (I do not like the term DSL, so I use builders instead). For example I desperately want something nice to replace `to` operation in tree builders, because `to` produces a lot of bugs due to confusion with pair production and `<<` would look nice in that role. But one need to understand that those operations will never be used as planned. For example if one introduces left shift operation with custom override, it will be used for a lot of things, but almost never for bitwise operations (because bitwise operations are rather rare). So one need to understand that when introducing something like this.

**Comment by philipguin** [comment]
I'm willing to take bitwise operators, with or without operator overloading. Arguments stating that it's "hard to read" can be made about literally every language feature, I don't find them to be particularly compelling. And in any case, I find these:

```
val flags = DISABLE_GRAVITY | ENABLE_CCD | IS_KINEMATIC
val colorRgb = (r / 255f).toInt() << 16 | (g / 255f).toInt() << 8 | (b / 255f).toInt() << 0
```
...way more legible than these:

```
val flags = DISABLE_GRAVITY or ENABLE_CCD or IS_KINEMATIC
val colorRgb = ((r / 255f).toInt() lsh 16) or ((g / 255f).toInt() lsh 8) or ((b / 255f).toInt() lsh 0)
```

Specifically because:

A) `or` implies "logical or," not "bitwise or," especially to someone new. It's actually even more confusing than just `|`, which a newcomer will at least either _know_ they don't know, or infer knowledge of from the other C-like languages. You'll notice that the first line (which is composed of actual PhysX flags) _reads_ as something entirely different from what it _means_.
B) Extra parentheses and relatively large 'operators' made of non-distinct characters remove clarity. The 'operators' in particular are far less visually identifiable at a glance.

Now, you can argue that the second line should be wrapped in a function anyways, but of course that function _also_ has to be read, and in system code, you'll be doing lots of twiddling of this sort that's not so easily or justifiably abstracted away.

I think the up-front cost of learning the operators is well worth the benefit to anyone who uses them, and is a cost of type that _everyone_ decides is worth paying when they first learn a programming language. (Not to mention that that cost is _less_ than what's currently being offered in Kotlin!) If you find it hard to read, then I'm not sure what forbidding these operators is doing for that issue.

**Comment by avosirenfal** [comment]
I often work with bitpacked data structures. You find them while encoding audio, video, and images, as well as in game file formats and network protocols.

The lack of bitwise operators makes porting existing code to Kotlin a huge headache. I love Kotlin, but this has personally ruled the language out for a large variety of things I'd otherwise like to use it for. Ditto with many of my colleagues.

The people arguing here that the operators are "rarely used" means "_I_ rarely have to use them". The fact that your area of programming doesn't need bitwise operations does not imply that is true for others.

I'm totally fine with overloading being disallowed, though I think the arguments against it are spurious. Language features in many languages can be abused to write unmanageable code, the onus is on us to write good software.

The mindset of "protecting people from confusing code" is what led to Java not having unsigned integers - a clear mistake that ultimately inhibited porting code from other languages, and interacting with structures produced by other languages. Do we really want to repeat that mistake again with a language that set out to fix the flaws in Java's language philosophy?

**Comment by Richyeoh** [comment]
> It would be nice:
> ```kotlin
> operator fun PrintStream.shl(any: Any) = apply { print(any) }
> 
> System.out << "hi!" << System.lineSeparator()
> ```

Some like c ++ language?

**Comment by Richyeoh** [comment]
> It would be nice:
> 
> ```kotlin
> operator fun PrintStream.shl(any: Any) = apply { print(any) }
> 
> System.out << "hi!" << System.lineSeparator()
> ```

Sorry, do I mean something like C++ syntax that?

**Comment by quickstep23** [comment]
Am I missing something, or have JetBrains silently introduced the possibility for infix operators?
Have a look at https://github.com/JetBrains/kotlin-spark-api/wiki/API-Reference

Functions
---

&& | infix fun Column.&&(c: Column): Column
-- | --
== | infix fun Column.==(c: Column): Column

**Comment by edrd-f** [comment]
[Not really](https://github.com/JetBrains/kotlin-spark-api/blob/e328d8da2843163e2931dc73f6fd22e782bcf84a/kotlin-spark-api-3.0.0_2.12/src/main/kotlin/org/jetbrains/kotlinx/spark/api/ApiV1.kt#L162). It's just that the backticks were omitted.

**Comment by Constellarise** [comment]
Something that might not have been considered: I cannot tell left from right without doing the L hand symbol. Never have been able to. I do quite a few bitshifty operations with my code, and I have a bug relating to shifting in the wrong direction that aren't obvious due to not being able to tell them apart nearly every single time I write a function with shifting. With the `<<` and `>>` operators, however, I don't need to remember if a shift is left or right; it's going in the direction of the arrows.

**Comment by Hc747** [comment]
Would love to see this KEEP accepted.

**Comment by elizarov** [comment]
Thanks a lot for a fruitful discussion in this KEEP. Having bitwise operators in the language definitely improves bit-manipulating code. However, it also adds complexity and creates additional potential for abuse. So far, I cannot say when or if Kotlin will have them.

In order to keep the discussion in one place, though, I'm closing this KEEP in favor of keeping track of this feature request in this issue: https://youtrack.jetbrains.com/issue/KT-1440
